README
#
# synapse: a small XMPP server
# README: at-a-glance documentation
#
# Copyright (c) 2006 Eric Will <rakaur@malkier.net>
#
# $Id$
#

= synapse
=== A Lightweight XMPP Server Implementation

This program is free but copyrighted software; see <tt>doc/license.txt</tt> for
details.

Information about +synapse+ may be found at <tt>http://synapse.malkier.net/</tt>.

== Credits

This program was written from scratch, based only off of my other Ruby programs,
none of which have ever been released. This program borrows some ideas, and
perhaps a small amount of code (mostly for IDN, TLS, and SASL) from the
netxmpp-ruby project: <tt>http://netxmpp-ruby.jabberstudio.org/</tt>.

Currently synapse is written largely by me:
  rakaur, Eric Will <rakaur@malkier.net>

With various code from:
  sycobuny, Stephen Belcher <sycobuny@malkier.net>

Thanks to "official" alpha/beta testers:
  naki, Adam Walls <naki@riacon.net>
  rintaun, Matt Lanigan <rintaun@projectxero.net>

And of course, thanks to all of the people who tested and posted bugs, etc.

Special thanks to:
  stpeter, Peter Saint-Andre <stpeter@jabber.org>
     for giving this project a chance, for putting up with my
     endless questions on the RFCs that you so kindly published,
     and of course for all the hard work you put into XMPP.

== References

  [   XMPP-CORE] -- RFC 3920 -- XMPP: Core
  [     XMPP-IM] -- RFC 3921 -- XMPP: Instant Messaging and Presence
  [      BASE64] -- RFC 3548 -- The Base16, Base32, and Base64 Data Encodings
  [  DIGEST-MD5] -- RFC 2831 -- Using Digest Authentication as a SASL Mechanism
  [       PLAIN] -- RFC 4616 -- The PLAIN SASL Mechanism
  [  STRINGPREP] -- RFC 3454 -- Preparation of Internationalized Strings
  [         IDN] -- RFC 3491 -- Stringprep Profile for Domain Names
  [        SASL] -- RFC 2222 -- Simple Authentication and Security Layer
  [         SRV] -- RFC 2782 -- DNS RR SRV
  [         TLS] -- RFC 2246 -- Transport Layer Security
  [  IQ-VERSION] -- XEP 0092 -- Software Verson
  [       DISCO] -- XEP 0030 -- Service Discovery
  [  MSGOFFLINE] -- XEP 0160 -- Best Practices for Handling Offline Messages
  [       DELAY] -- XEP 0203 -- Delayed Delivery
  [ IQ-REGISTER] -- XEP 0077 -- In-Band Registration
  [  VCARD-TEMP] -- XEP 0054 -- vcard-temp
  [VCARD-AVATAR] -- XEP 0153 -- vCard-Based Avatar


== Presentation

+synapse+ is a small, lightweight, strictly RFC-compliant Jabber/XMPP server.
It doesn't require a system-wide install, so that XMPP can hopefully be
brought to the world of shells. As part of this goal, +synapse+ doesn't rely
on many processes running concurrently (+jabberd+), and doesn't rely on
an exotic language to be installed (+ejabberd+) (unless you consider Ruby
exotic, I guess, but it's installed in all of *my* shells).

+synapse+ should happily interoperate with any server that is either RFC
compliant, or supports, at the very least, the Dialback protocol. +synapse+
is quite strict in regards to enforcing the RFC, with one exception for
legacy +jabberd+ servers that don't properly specify the <tt>jabber:server</tt>
namespace.

== Installation

+synapse+ comes pretty much installed, as there's no compile-time.

=== Dependencies:

* ruby (>= 1.8.0)
* openssl
* idn (>= 0.0.1) (rubygem: <tt>gem install --remote idn</tt>)
* rake (rubygem: <tt>gem install --remote rake</tt>)

We require ruby for obvious reasons.

We require OpenSSL for TLS streams.

We require IDN for Nameprep. Thankfully, the RubyGem 'idn' already includes
support for Nodeprep and Resourceprep.

We require rake to run the tests and generate documentation.

Once all dependencies are met, edit <tt>etc/xmppd.conf</tt> in accordance to
the example in <tt>etc/example.conf</tt>. Once the configuration is to your
liking, you may of course move the installation to wherever you like, and run
the main program with <tt>./bin/xmppd</tt>.

== Contact

If you have a bug, support request, patch, or feature request, please use
our tracker: <tt>http://synapse.malkier.net/</tt>.

Also, my JID is <tt>rakaur@malkier.net</tt>. Keep in mind if you message me
about a bug, support request, patch, or feature request, I'm just going to point
you to the tracker. If I handled everyone's problems on a case-by-case
basis I'd certainly be crazier than I already am.

There is a mailing list, xmppd-dev@lists.malkier.net.

If you've read this far, congratulations. You are among the few elite people
that actually read documentation. Thank you.


ChangeLog
commit 6d8bea847fc851af460e19ba0271d00416caef91
Author: Eric Will <rakaur@malkier.net>
Date:   Tue Nov 4 09:10:56 2008 -0500

    Implement an XML parser using expat.
    
    This tries to use expat first, then falls back to REXML.
    There is a major difference: REXML will detect not-well-formed XML, and still manage
    to patch incomplete stanzas together okay. expat will never recognize a bad stanza, but
    instead will wait for the rest of it to be valid or just sit there until the buffer explodes.

 lib/xmppd/xmpp/parser.rb       |  150 +--------------------------------------
 lib/xmppd/xmpp/parser_expat.rb |  133 +++++++++++++++++++++++++++++++++++
 lib/xmppd/xmpp/parser_rexml.rb |  152 ++++++++++++++++++++++++++++++++++++++++
 3 files changed, 289 insertions(+), 146 deletions(-)

commit 8d35ec24282011f8aaed429fd77dbf3b61f1093f
Author: Eric Will <rakaur@malkier.net>
Date:   Mon Nov 3 07:31:53 2008 -0500

    Make a note about the RFC being impossible.

 ChangeLog                |   36 ++++++++++++++++++++++++++++++++++++
 Rakefile                 |    3 ++-
 lib/xmppd/xmpp/parser.rb |    9 +++++----
 3 files changed, 43 insertions(+), 5 deletions(-)

commit 160be58ef40c47da8f6575997af2896d4e80ce97
Author: Eric Will <rakaur@malkier.net>
Date:   Sat Nov 1 16:22:26 2008 -0400

    Fixed a timer issue.
    
    Moving the timer stuff to after the fork(), as this kills the timer threads. Thanks Sho.
    Silly Ruby, threads are for real languages.

 lib/xmppd.rb |   22 +++++++++++-----------
 1 files changed, 11 insertions(+), 11 deletions(-)

commit 537f9a377499865e9fe7418761c95e798c156c62
Author: Eric Will <rakaur@malkier.net>
Date:   Fri Oct 31 15:11:57 2008 -0400

    Fix some things up for git.
    
    Updated Rakefile to work with git instead of svn.
    Changed the ChangeLog from svn log to git log.

 ChangeLog            | 2642 +++++++++++++++++++++++++++-----------------------
 Rakefile             |   26 +-
 lib/xmppd/version.rb |    6 +-
 3 files changed, 1440 insertions(+), 1234 deletions(-)

commit eafa62403543b12e1917f8bd9e3ffde6212de29f
Author: Eric Will <rakaur@malkier.net>
Date:   Fri Oct 31 14:09:41 2008 -0400

    Added a copy of the BSD license, closes #83.

 README          |    2 +-
 doc/license.txt |   27 +++++++++++++++++++++++++++
 2 files changed, 28 insertions(+), 1 deletions(-)

commit 4cf0e07fe7acb046a4b52726c2267453584ca5b3
Author: Eric Will <rakaur@malkier.net>
Date:   Sat Oct 25 13:01:59 2008 -0400

    Working on presence subscriptions.
    
    I'm pretty sure this is mostly done, but I wanna make sure before I merge back.

 .gitignore                 |    7 ++
 lib/xmppd/db.rb            |    7 +-
 lib/xmppd/xmpp/iq_get.rb   |    2 +-
 lib/xmppd/xmpp/iq_set.rb   |   21 +++-
 lib/xmppd/xmpp/presence.rb |  240 +++++++++++++++++++++++++++++++++++++++++++-
 5 files changed, 263 insertions(+), 14 deletions(-)

commit a0114b5430f69cccbfba34262a3174a216427916
Author: Eric Will <rakaur@malkier.net>
Date:   Sun Oct 19 04:46:48 2008 +0000

    Added some new docs.
    Start working on the rest of subscriptions.
    Fix a few bugs.
    Disallow multiple resource binding to one stream.
    Start working on last-activity.
    Track uptime.
    The Parser.dispatch method will soon be very critical. See #60.

 ChangeLog                         |   82 +++++++++++
 README                            |   33 +++--
 doc/extensions/html/xep-0153.html |  251 ++++++++++++++++++++++++++++++++
 doc/extensions/xml/xep-0153.xml   |  284 +++++++++++++++++++++++++++++++++++++
 lib/xmppd/db.rb                   |    4 +-
 lib/xmppd/var.rb                  |    3 +
 lib/xmppd/xmpp/iq_set.rb          |    9 +-
 lib/xmppd/xmpp/parser.rb          |    3 +-
 lib/xmppd/xmpp/presence.rb        |   51 +++++++
 9 files changed, 700 insertions(+), 20 deletions(-)

commit 45854ae7b6e6bdc03f751d77031d6dabbc802d8f
Author: Eric Will <rakaur@malkier.net>
Date:   Sat Oct 18 20:37:19 2008 +0000

    Just updating some documentation.

 doc/drafts/draft-saintandre-rfc3920bis-08.html     | 7220 ---------------
 doc/drafts/draft-saintandre-rfc3920bis-08.txt      | 9240 --------------------
 doc/drafts/draft-saintandre-rfc3921bis-07.html     | 4122 ---------
 doc/drafts/draft-saintandre-rfc3921bis-07.txt      | 5376 ------------
 .../html/draft-saintandre-rfc3920bis-08.html       | 7220 +++++++++++++++
 .../html/draft-saintandre-rfc3921bis-07.html       | 4122 +++++++++
 doc/drafts/text/draft-saintandre-rfc3920bis-08.txt | 9240 ++++++++++++++++++++
 doc/drafts/text/draft-saintandre-rfc3921bis-07.txt | 5376 ++++++++++++
 8 files changed, 25958 insertions(+), 25958 deletions(-)

commit 1b3ea5276d4f8ec9ffc3dcbf1e891a1174f575b8
Author: Eric Will <rakaur@malkier.net>
Date:   Sat Oct 18 20:35:15 2008 +0000

    Just updating some documentation.

 doc/drafts/draft-saintandre-rfc3920bis-07.html    | 7214 ----------------
 doc/drafts/draft-saintandre-rfc3920bis-07.txt     | 9240 ---------------------
 doc/drafts/draft-saintandre-rfc3920bis-08.html    | 7220 ++++++++++++++++
 doc/drafts/draft-saintandre-rfc3920bis-08.txt     | 9240 +++++++++++++++++++++
 doc/drafts/draft-saintandre-rfc3921bis-06.html    | 4122 ---------
 doc/drafts/draft-saintandre-rfc3921bis-06.txt     | 5376 ------------
 doc/drafts/draft-saintandre-rfc3921bis-07.html    | 4122 +++++++++
 doc/drafts/draft-saintandre-rfc3921bis-07.txt     | 5376 ++++++++++++
 doc/drafts/xml/draft-saintandre-rfc3920bis-08.xml | 6047 ++++++++++++++
 doc/drafts/xml/draft-saintandre-rfc3921bis-07.xml | 3525 ++++++++
 doc/extensions/html/xep-0030.html                 |  808 ++
 doc/extensions/html/xep-0054.html                 |  618 ++
 doc/extensions/html/xep-0077.html                 |  817 ++
 doc/extensions/html/xep-0092.html                 |  123 +
 doc/extensions/html/xep-0160.html                 |  107 +
 doc/extensions/html/xep-0186.html                 |  169 +
 doc/extensions/html/xep-0203.html                 |  149 +

lib/xmppd/db.rb
#
# synapse: a small XMPP server
# db.rb: database classes
#
# Copyright (c) 2006 Eric Will <rakaur@malkier.net>
#
# $Id: db.rb 90 2008-10-19 04:46:48Z rakaur $
#

#
# Import required Ruby modules.
#
require 'digest/md5'
require 'idn'
require 'rexml/document'
require 'yaml'

#
# Import required xmppd modules.
#
require 'xmppd/xmpp/stanza'

#
# The DB namespace.
#
module DB

extend self

#
# Database exception.
#
class DBError < Exception
end

#
# Represents a registered user.
#
class User
    @@users = {}
    @@need_dump = false

    attr_accessor :last, :offline_stanzas, :vcard
    attr_reader   :node, :domain, :password, :resources, :roster

    def initialize(node, domain, password)
        @resources = {}
        @roster = {}
        @offline_stanzas = []

        @node = IDN::Stringprep.nodeprep(node[0, 1023])

        if domain =~ /^(\d+)\.(\d+)\.(\d+)\.(\d+)$/
            @domain = domain
        else
            @domain = IDN::Stringprep.nameprep(domain[0, 1023])
        end

        unless $config.hosts.include?(@domain)
            raise DBError, "we do not serve host: #{@domain}"
        end

        @password = "%s:%s:%s" % [@node, @domain, password]
        @password = Digest::MD5.digest(@password)

        raise DBError, "#{jid} already exists" if @@users[jid]

        @@users[jid] = self

        $log.xmppd.info "new user: #{jid}"

        @@need_dump = true
    end

    ######
    public
    ######

    def to_yaml_properties
        %w( @node @domain @password @roster @last @offline_stanzas @vcard )
    end

    def User.users
        @@users
    end

    def User.need_dump?
        true #@@need_dump -- XXX figure out offline_stanzas
    end

    def User.load
        begin
            File.open('var/db/users.yaml') do |f|
                @@users = YAML.load(f)
                @@users ||= {}
            end
        rescue Exception => e
            puts "xmppd: failed to load user database: #{e}"
            exit
        else
            $log.xmppd.info "loaded #{@@users.length} users"
        end
    end

    def User.dump
        return unless need_dump?

        # Prune away empty entries.
        @@users.delete_if { |k, v| v.nil? } if @@users.has_value?(nil)

        @@users.each do |uk, uv|
            uv.roster.delete_if { |k, v| v.nil? } if uv.roster.has_value?(nil)
        end

        begin
            Dir.mkdir('var/db') unless File.exists?('var/db')

            File.open('var/db/users.yaml', 'w') { |f| YAML.dump(@@users, f) }
        rescue Exception => e
            $log.xmppd.warn "failed to write user database: #{e}"
            return false
        else
            $log.xmppd.info "wrote #{@@users.length} users"
            @@need_dump = false
            return true
        end
    end

    def User.auth(jid, password, plain = false)
        return false unless @@users[jid]

        if plain
            node, domain = jid.split('@')
            check = "%s:%s:%s" % [node, domain, password]
            check = Digest::MD5.digest(check)

            return true if check == @@users[jid].password
        else
            return true if password == @@users[jid].password
        end

        return false
    end

    def User.delete(jid)
        raise DBError, "#{jid} does not exist" unless @@users[jid]

        user = @@users[jid]
        @@users.delete(jid)

        # Disconnect any active resources.
        user.resources.each { |n, rec| rec.stream.close } if user.resources

        # XXX - remove them from rosters?
        user.roster.each do |c|
        end

        $log.xmppd.info "user delete: #{jid}"

        @@need_dump = true
    end

    def jid
        @node + '@' + @domain
    end

    def operator?
        m = $config.operator.find { |oper| oper.jid == jid }
        return false unless m
        return m
    end

    def password=(newpass)
        newpass = "%s:%s:%s" % [@node, @domain, newpass]
        @password = Digest::MD5.digest(newpass)

        $log.xmppd.info 'password change for %s' % jid

        @@need_dump = true
    end

    def add_contact(contact)
        unless contact.kind_of?(Contact)
            raise DBError, "contact isn't a Contact class"
        end

        @roster[contact.jid] = contact

        $log.xmppd.debug "DB::User.add_contact(): #{jid} -> #{contact.jid}"

        @@need_dump = true
    end

    def delete_contact(ujid)
        if @roster[ujid]
            @roster.delete(ujid)

            $log.xmppd.debug "DB::User.delete_contact(): #{jid} -> #{ujid}"

            @@need_dump = true
        end
    end

    def add_resource(resource)
        unless resource.class == XMPP::Client::Resource
            raise DBError, "resource isn't a Resource class"
        end

        @resources ||= {}
        @resources[resource.name] = resource
    end

    def delete_resource(resource)
        unless resource.class == XMPP::Client::Resource
            raise DBError, "resource isn't a Resource class"
        end

        @resources.delete_if { |j, rec| rec == resource }
    end

    # True if we have any connected resources.
    def available?
        $-w = false
        if @resources.nil? or @resources.empty?
            return false
        else
            return true
        end
        $-w = true
    end

    # True if we're subscribed to their presence.
    def subscribed?(user)
        return true if user == self

        myc = @roster[user.jid]
        return false unless myc

        return true if myc.subscription == 'to'
        return true if myc.subscription == 'both'

        return false
    end

    def offline_stanza(stanza, myhost)
        # This implements XEP-0203.
        dt = Time.now.utc.strftime("%Y-%m-%dT%H:%M:%SZ")
        delay = REXML::Element.new('delay')
        delay.add_attribute('stamp', dt)
        delay.add_attribute('from', myhost)
        delay.add_namespace('urn:xmpp:delay')
        delay.text = 'Offline Storage'

        stanza << delay
        @offline_stanzas << stanza.to_s
    end

    #
    # Return the resource with the highest priority.
    #
    def front_resource
        recs = {}

        # If there is a tie, the last one in the loop wins.
        @resources.each do |name, rec|
            next unless rec.available?
            next unless rec.presence_stanza.elements['priority']
            p = rec.presence_stanza.elements['priority'].text.to_i
            next if p < 0 # Skip negative priorities.

            recs[p] = rec
        end

        return recs[recs.keys.max]
    end

    #
    # Returns an array of Contacts in our roster that
    # are subscribed to our presence.
    #
    def roster_subscribed_from
        @roster.find_all { |j, contact| contact.user.subscribed?(self) }
    end

    #
    # Returns an array of Contacts in our roster that
    # we are subscribed to.
    def roster_subscribed_to
        @roster.find_all { |j, contact| subscribed?(contact) }
    end

    def roster_to_xml
        query = XMPP::Stanza.new_query('jabber:iq:roster')
        @roster.each { |name, contact| query << contact.to_xml }
        return query
    end

    def clean_roster
        return unless @roster
        return if @roster.empty?

        @roster.each do |j, c|
            if c.subscription == 'none' and c.pending_none?
                delete_contact(j)

                if available?
                    #
                    # God, Ruby's threads are totally useless. I can't even
                    # get a thread to look at stanza.rb because *something*
                    # happens and everything stops working.
                    #
                    iq    = REXML::Element.new('iq')
                    query = REXML::Element.new('query')
                    item  = REXML::Element.new('item')

                    iq.add_attribute('id', 'rubysucks')
                    iq.add_attribute('type', 'set')

                    query.add_namespace('jabber:iq:roster')

                    item.add_attribute('jid', j)
                    item.add_attribute('subscription', 'remove')

                    query << item
                    iq    << query

                    @resources.each { |n, rec| rec.stream.write iq }
                end
            end

            next unless c.pending_out?

            if ($time - c.stime) >= 86400 # One day.
                presence = REXML::Element.new('presence')
                presence.add_attribute('type', 'subscribe')
                presence.add_attribute('to', j)
                presence.add_attribute('from', jid)

                if available?
                    front_resource.stream.process_stanza(presence)
                    c.stime = $time
                end
            end
        end
    end
end

#
# A contact in a roster.
#
class Contact
    attr_accessor :groups, :name, :pending, :stime
    attr_reader   :subscription

    PEND_NONE = 0x00000000
    PEND_IN   = 0x00000001
    PEND_OUT  = 0x00000002

    def initialize
        @groups = []
        @name = nil
        @subscription = 'none'
        @pending = PEND_NONE
    end

    ######
    public
    ######

    def subscription=(value)
        unless value =~ /^(to|from|both|none)$/
            raise ArgumentError, 'subscription must be /(to|from|both|none)/'
        end

        @subscription = value
    end

    def pending_in=(value)
        unless value == true or value == false
            raise ArgumentError, 'value must be true or false'
        end

        if value
            @pending |= PEND_IN
        else
            @pending &= ~PEND_IN
        end
    end

    def pending_out=(value)
        unless value == true or value == false
            raise ArgumentError, 'value must be true or false'
        end

        if value 
            @pending |= PEND_OUT
        else
            @pending &= ~PEND_OUT
        end
    end

    def pending_none=(value)
        unless value == true or value == false
            raise ArgumentError, 'value must be true'
        end

        @pending = PEND_NONE
    end

    def pending_none?
        return (@pending == PEND_NONE) ? true : false
    end

    def pending_in?
        return (PEND_IN & @pending != 0) ? true : false
    end

    def pending_out?
        return (PEND_OUT & @pending != 0) ? true : false
    end

    def to_xml
        item = REXML::Element.new('item')
        item.add_attribute('jid', self.jid)
        item.add_attribute('name', @name) if @name
        item.add_attribute('subscription', @subscription)

        item.add_attribute('ask', 'subscribe') if pending_out?

        @groups.each do |g|
            group = REXML::Element.new('group')
            group.text = g
            item << group
        end if @groups

        item
    end
end

class LocalContact < Contact
    attr_reader :user

    def initialize(user)
        super()

        raise DBError, 'user is not a DB::User' unless user.class == User

        @user = user
    end

    ######
    public
    ######

    def to_yaml_properties
        %w( @user @stime @subscription @pending @name @groups )
    end

    def jid
        @user.jid
    end
end

class RemoteContact < Contact
    attr_reader :node, :domain, :resources

    def initialize(node, domain, resource = nil)
        super()

        @resources = []
        @node = IDN::Stringprep.nodeprep(node)
        @domain = IDN::Stringprep.nameprep(domain)

        add_resource(resource) if resource
    end

    ######
    public
    ######

    def to_yaml_properties
        %w( @node @stime @domain @subscription @pending @name @groups )
    end

    def jid
        @node + '@' + @domain
    end

    def add_resource(resource)
        @resources << IDN::Stringprep.resourceprep(resource)
    end

    def delete_resource(resource)
        @resources.delete_if { |r| r == resource }
    end
end

end # module DB
